File: runtime/src/direwolf/wire/envelope.py
"""The common envelope and the three family decoders.

Mirrors ``dwk-proto``'s ``envelope``, ``dwkp``, ``dwcp`` and ``events`` modules,
in the same decode order, so both implementations report the same failure for
the same single-fault input:

1. the document is an object;
2. ``v`` present, an integer, in range, and supported;
3. ``type``, ``schema``, ``schema_version``;
4. family lookup (DWKP rejects unknown operations; DWCP and events retain them);
5. undeclared envelope members (DWKP rejects; DWCP and events preserve);
6. ``id``, ``ts``, then the optional fields against the message's rules;
7. ``epoch`` requires ``session_id``; the ``id`` prefix matches the type;
8. ``payload`` present and an object, then decoded by its generated type.

The per-message rules and payload types are generated into
:mod:`direwolf.proto` and passed in; this module holds only the mechanism. The
field formats below are checked against ``schemas/common/envelope.v1.schema.json``
by the test suite.
"""

from __future__ import annotations

from collections.abc import Mapping
from dataclasses import dataclass, field
from typing import Final, Literal, Protocol, Self

from direwolf.wire import frame, jcs, strict_json, validate
from direwolf.wire.errors import (
    FORBIDDEN_FIELD,
    FRAME_TOO_LARGE,
    INCONSISTENT,
    INVALID_FORMAT,
    MAX_MESSAGE_BYTES,
    MAX_SAFE_INTEGER,
    MISSING_FIELD,
    NULL_NOT_ALLOWED,
    OUT_OF_RANGE,
    SUPPORTED_ENVELOPE,
    UNKNOWN_OPERATION,
    VERSION_UNSUPPORTED,
    WRONG_TYPE,
    ProtocolError,
    VersionRange,
    schema_violation,
)
from direwolf.wire.strict_json import JsonValue
from direwolf.wire.validate import Cx

type Presence = Literal["required", "optional", "forbidden"]

ENVELOPE_KEYS: Final = (
    "v",
    "id",
    "type",
    "schema",
    "schema_version",
    "ts",
    "correlation_id",
    "causation_id",
    "session_id",
    "run_id",
    "epoch",
    "idempotency_key",
    "payload",
)

MESSAGE_TYPES: Final = ("request", "response", "event")
ID_PREFIX: Final = {"request": "msg", "response": "msg", "event": "evt"}

SCHEMA_NAME_PATTERN: Final = "^direwolf(\\.[a-z][a-z0-9_]{0,31}){1,6}$"
TIMESTAMP_PATTERN: Final = "^[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}\\.[0-9]{3}Z$"
IDEMPOTENCY_KEY_PATTERN: Final = "^[A-Za-z0-9][A-Za-z0-9._:-]{0,127}$"

_VERSION = validate.integer(1, 65535)
_TYPE = validate.enumeration(MESSAGE_TYPES)
_SCHEMA = validate.text(128, SCHEMA_NAME_PATTERN, None)
_TS = validate.text(24, TIMESTAMP_PATTERN, "rfc3339-utc-millis")
_ANY_ID = validate.identifier(None)
_OPTIONAL_CHECKS: Final[dict[str, validate.Check[str] | validate.Check[int]]] = {
    "correlation_id": _ANY_ID,
    "causation_id": _ANY_ID,
    "session_id": validate.identifier("ses"),
    "run_id": validate.identifier("run"),
    "epoch": validate.integer(1, MAX_SAFE_INTEGER),
    "idempotency_key": validate.text(128, IDEMPOTENCY_KEY_PATTERN, None),
}


class Payload(Protocol):
    """What every generated payload type provides."""

    @classmethod
    def decode(cls, value: JsonValue, cx: Cx) -> Self: ...

    def encode(self) -> dict[str, JsonValue]: ...


@dataclass(frozen=True, slots=True)
class Rules:
    """Presence of the optional envelope fields for one message."""

    correlation_id: Presence
    causation_id: Presence
    session_id: Presence
    run_id: Presence
    epoch: Presence
    idempotency_key: Presence

    def items(self) -> tuple[tuple[str, Presence], ...]:
        return (
            ("correlation_id", self.correlation_id),
            ("causation_id", self.causation_id),
            ("session_id", self.session_id),
            ("run_id", self.run_id),
            ("epoch", self.epoch),
            ("idempotency_key", self.idempotency_key),
        )


PERMISSIVE: Final = Rules("optional", "optional", "optional", "optional", "optional", "optional")


@dataclass(frozen=True, slots=True)
class MessageSpec:
    """One message this build knows, as generated from its schema."""

    schema: str
    message_type: str
    versions: VersionRange
    rules: Rules
    payload: type[Payload]


type Registry = Mapping[tuple[str, str], MessageSpec]


@dataclass(frozen=True, slots=True)
class Header:
    """The decoded envelope. Its fields are claims, not facts (ADR-0028)."""

    v: int
    id: str
    type: str
    schema: str
    schema_version: int
    ts: str
    correlation_id: str | None = None
    causation_id: str | None = None
    session_id: str | None = None
    run_id: str | None = None
    epoch: int | None = None
    idempotency_key: str | None = None

    def encode(
        self, payload: JsonValue, extensions: Mapping[str, JsonValue]
    ) -> dict[str, JsonValue]:
        out: dict[str, JsonValue] = {
            "v": self.v,
            "id": self.id,
            "type": self.type,
            "schema": self.schema,
            "schema_version": self.schema_version,
            "ts": self.ts,
        }
        for name in (
            "correlation_id",
            "causation_id",
            "session_id",
            "run_id",
            "epoch",
            "idempotency_key",
        ):
            value = getattr(self, name)
            if value is not None:
                out[name] = value
        out["payload"] = payload
        for key, value in extensions.items():
            out[key] = value
        return out


@dataclass(frozen=True, slots=True)
class Preamble:
    v: int
    type: str
    schema: str
    schema_version: int


def read_preamble(value: JsonValue, supported: VersionRange = SUPPORTED_ENVELOPE) -> Preamble:
    if not isinstance(value, dict):
        raise schema_violation(
            WRONG_TYPE, "", f"a message is an object, found {validate.type_name(value)}"
        )
    v = _peek(value, "v", _VERSION)
    if not supported.contains(v):
        raise ProtocolError(
            VERSION_UNSUPPORTED,
            f"envelope version {v} is not supported",
            violation=OUT_OF_RANGE,
            path="/v",
            supported=supported,
        )
    message_type = _peek(value, "type", _TYPE)
    schema = _peek(value, "schema", _SCHEMA)
    schema_version = _peek(value, "schema_version", _VERSION)
    return Preamble(v, message_type, schema, schema_version)


def _peek[T](obj: dict[str, JsonValue], name: str, check: validate.Check[T]) -> T:
    if name not in obj:
        raise schema_violation(MISSING_FIELD, f"/{name}", f"missing envelope field {name!r}")
    cx = Cx()
    cx.push(name)
    raw = obj[name]
    if raw is None:
        raise cx.violation(NULL_NOT_ALLOWED, "null is not a value")
    return check(raw, cx)


def decode_header(
    obj: dict[str, JsonValue], preamble: Preamble, rules: Rules, preserve: bool
) -> tuple[Header, dict[str, JsonValue], dict[str, JsonValue]]:
    cx = Cx()
    known, extensions = validate.partition(obj, ENVELOPE_KEYS, preserve, cx)
    identifier = validate.take_required(known, "id", cx, _ANY_ID)
    ts = validate.take_required(known, "ts", cx, _TS)
    optional: dict[str, JsonValue] = {}
    for name, presence in rules.items():
        check = _OPTIONAL_CHECKS[name]
        if presence == "forbidden":
            if name in known:
                raise schema_violation(
                    FORBIDDEN_FIELD, f"/{name}", f"envelope field {name!r} is not permitted"
                )
        elif presence == "required":
            optional[name] = validate.take_required(known, name, cx, check)
        else:
            optional[name] = validate.take_optional(known, name, cx, check)
    if optional.get("epoch") is not None and optional.get("session_id") is None:
        raise schema_violation(INCONSISTENT, "/epoch", "session_id is required with epoch")
    if identifier.partition("_")[0] != ID_PREFIX[preamble.type]:
        raise schema_violation(
            INVALID_FORMAT, "/id", f"a {preamble.type} uses a {ID_PREFIX[preamble.type]}_ id"
        )
    if "payload" not in known:
        raise schema_violation(MISSING_FIELD, "/payload", "missing envelope field 'payload'")
    payload = known["payload"]
    if payload is None:
        raise schema_violation(NULL_NOT_ALLOWED, "/payload", "null is not a value")
    if not isinstance(payload, dict):
        raise schema_violation(
            WRONG_TYPE, "/payload", f"expected object, found {validate.type_name(payload)}"
        )

    def opt_str(name: str) -> str | None:
        value = optional.get(name)
        return value if isinstance(value, str) else None

    epoch = optional.get("epoch")
    header = Header(
        v=preamble.v,
        id=identifier,
        type=preamble.type,
        schema=preamble.schema,
        schema_version=preamble.schema_version,
        ts=ts,
        correlation_id=opt_str("correlation_id"),
        causation_id=opt_str("causation_id"),
        session_id=opt_str("session_id"),
        run_id=opt_str("run_id"),
        epoch=epoch if isinstance(epoch, int) else None,
        idempotency_key=opt_str("idempotency_key"),
    )
    return header, payload, extensions


def _decode_payload(spec: MessageSpec, payload: dict[str, JsonValue]) -> Payload:
    cx = Cx()
    cx.push("payload")
    return spec.payload.decode(payload, cx)


# ---- DWKP: strict ------------------------------------------------------------


@dataclass(frozen=True, slots=True)
class DwkpMessage:
    header: Header
    body: Payload


def decode_dwkp(data: bytes, registry: Registry) -> DwkpMessage:
    """Decode a DWKP frame body. Rejects everything ``dwk-proto`` rejects."""
    value = strict_json.parse(data, strict_json.SAFE_INTEGER)
    preamble = read_preamble(value)
    spec = registry.get((preamble.type, preamble.schema))
    if spec is None:
        raise ProtocolError(
            UNKNOWN_OPERATION,
            f"no DWKP {preamble.type} named {preamble.schema} exists in this build",
            path="/schema",
        )
    if not spec.versions.contains(preamble.schema_version):
        raise ProtocolError(
            VERSION_UNSUPPORTED,
            f"{spec.schema} schema_version {preamble.schema_version} is not supported",
            violation=OUT_OF_RANGE,
            path="/schema_version",
            supported=spec.versions,
        )
    if not isinstance(value, dict):  # read_preamble has already rejected this
        raise schema_violation(WRONG_TYPE, "", "a message is an object")
    header, payload, _ = decode_header(value, preamble, spec.rules, preserve=False)
    return DwkpMessage(header, _decode_payload(spec, payload))


def encode_dwkp(message: DwkpMessage, registry: Registry) -> bytes:
    """Canonical bytes, re-decoded before return: never emit what would be rejected."""
    spec = registry.get((message.header.type, message.header.schema))
    if spec is None or not isinstance(message.body, spec.payload):
        raise schema_violation(
            INCONSISTENT, "/schema", "the header's type and schema do not match the body"
        )
    data = jcs.canonicalize(message.header.encode(message.body.encode(), {}))
    if decode_dwkp(data, registry) != message:
        raise schema_violation(INCONSISTENT, "", "message does not survive its own round trip")
    return data


def frame_dwkp(message: DwkpMessage, registry: Registry) -> bytes:
    return frame.encode(encode_dwkp(message, registry))


# ---- DWCP: forward-compatible ----------------------------------------------------


@dataclass(frozen=True, slots=True)
class DwcpMessage:
    header: Header
    header_extensions: dict[str, JsonValue] = field(default_factory=dict)
    body: Payload | None = None
    unknown_payload: dict[str, JsonValue] | None = None


def decode_dwcp(data: bytes, registry: Registry) -> DwcpMessage:
    """Decode a DWCP message, preserving what this build does not declare."""
    if len(data) > MAX_MESSAGE_BYTES:
        raise ProtocolError(FRAME_TOO_LARGE, f"message of {len(data)} bytes exceeds the limit")
    value = strict_json.parse(data, strict_json.IJSON)
    preamble = read_preamble(value)
    if not isinstance(value, dict):  # read_preamble has already rejected this
        raise schema_violation(WRONG_TYPE, "", "a message is an object")
    spec = registry.get((preamble.type, preamble.schema))
    if spec is not None and spec.versions.contains(preamble.schema_version):
        header, payload, extensions = decode_header(value, preamble, spec.rules, preserve=True)
        return DwcpMessage(header, extensions, _decode_payload(spec, payload))
    header, payload, extensions = decode_header(value, preamble, PERMISSIVE, preserve=True)
    return DwcpMessage(header, extensions, None, payload)


def encode_dwcp(message: DwcpMessage) -> bytes:
    if message.body is not None:
        payload: JsonValue = message.body.encode()
    else:
        payload = dict(message.unknown_payload or {})
    return jcs.canonicalize(message.header.encode(payload, message.header_extensions))


# ---- event records: retained verbatim ---------------------------------------------


@dataclass(frozen=True, slots=True)
class KnownEvent:
    header: Header
    header_extensions: dict[str, JsonValue]
    event: Payload


@dataclass(frozen=True, slots=True)
class UnknownEvent:
    schema: str | None
    schema_version: int | None
    reason: ProtocolError


@dataclass(frozen=True, slots=True)
class InvalidEvent:
    error: ProtocolError


@dataclass(frozen=True, slots=True)
class EventRecord:
    """A record's exact bytes and how far they were interpreted."""

    raw: bytes
    view: KnownEvent | UnknownEvent | InvalidEvent


def read_event(data: bytes, registry: Registry) -> EventRecord:
    """Read a record. Raises only if the bytes are not a record at all."""
    if len(data) > MAX_MESSAGE_BYTES:
        raise ProtocolError(FRAME_TOO_LARGE, f"record of {len(data)} bytes exceeds the limit")
    value = strict_json.parse(data, strict_json.IJSON)
    if not isinstance(value, dict):
        raise schema_violation(WRONG_TYPE, "", "an event record is an object")
    schema = value.get("schema")
    schema_version = value.get("schema_version")
    loose_schema = schema if isinstance(schema, str) else None
    loose_version = (
        schema_version
        if isinstance(schema_version, int)
        and not isinstance(schema_version, bool)
        and 0 <= schema_version <= 65535
        else None
    )
    try:
        preamble = read_preamble(value)
    except ProtocolError as reason:
        return EventRecord(data, UnknownEvent(loose_schema, loose_version, reason))
    if preamble.type != "event":
        error = schema_violation("UNKNOWN_VARIANT", "/type", 'an event record has type "event"')
        return EventRecord(data, InvalidEvent(error))
    spec = registry.get(("event", preamble.schema))
    if spec is None or not spec.versions.contains(preamble.schema_version):
        unknown = ProtocolError(
            UNKNOWN_OPERATION, "unknown event schema and version", path="/schema"
        )
        return EventRecord(data, UnknownEvent(loose_schema, loose_version, unknown))
    try:
        header, payload, extensions = decode_header(value, preamble, spec.rules, preserve=True)
        event = _decode_payload(spec, payload)
    except ProtocolError as error:
        return EventRecord(data, InvalidEvent(error))
    return EventRecord(data, KnownEvent(header, extensions, event))

File: crates/dwk-proto/src/envelope.rs
//! The common envelope (`PROTOCOL.md` §1).
//!
//! Every message in every family — DWKP, DWCP and event records — has the same
//! envelope. What differs between families is the policy applied to it
//! (ADR-0023); what differs between operations is which optional envelope
//! fields are required, permitted or forbidden, declared per message in
//! [`EnvelopeRules`].
//!
//! # Decode order
//!
//! Fixed, so that every implementation reports the same failure for the same
//! input:
//!
//! 1. The document is an object.
//! 2. `v` — present, an integer, in range, and **supported**. Checked before
//!    anything else because a newer envelope may legitimately contain fields
//!    this receiver does not know; the right answer to such a message is
//!    `PROTOCOL_VERSION_UNSUPPORTED`, not `UNKNOWN_FIELD`.
//! 3. `type`, then `schema`, then `schema_version` — enough to identify the
//!    message. The family then looks it up.
//! 4. Undeclared envelope members (rejected or preserved by family).
//! 5. `id`, `ts`, `correlation_id`, `causation_id`, `session_id`, `run_id`,
//!    `epoch`, `idempotency_key`, each against the message's presence rule.
//! 6. Cross-field rules: an `epoch` requires a `session_id`; the `id` prefix
//!    matches the message type.
//! 7. `payload` — present and an object. Its contents are the family's concern.
//!
//! # Ownership of envelope fields
//!
//! The envelope carries **claims, not facts**. `ts` is the sender's clock.
//! `session_id`, `run_id` and `epoch` name authority the kernel issued; the
//! kernel checks them against `kernel.db` and never believes them
//! (ADR-0028, `PROTOCOL.md` §3). No envelope field is a policy input.

use crate::error::{ProtocolError, Violation, quote_key};
use crate::json::{Object, Value};
use crate::schema::Defs;
use crate::version::VersionRange;
use crate::wire::id::{AnyId, RunId, SessionId};
use crate::wire::scalar::{Epoch, IdempotencyKey, SchemaName, Timestamp, Version, wire_enum};
use crate::wire::{Cx, UnknownFields, WireType, partition_members, take_optional, take_required};

wire_enum! {
    /// What kind of message this is.
    MessageType {
        /// Sent by an initiator, expecting a response.
        Request = "request",
        /// Answers a request; carries the request id as `causation_id`.
        Response = "response",
        /// Records that something happened.
        Event = "event",
    }
}

impl MessageType {
    /// The identifier prefix messages of this type use for `id`.
    #[must_use]
    pub const fn id_prefix(self) -> &'static str {
        match self {
            Self::Request | Self::Response => "msg",
            Self::Event => "evt",
        }
    }
}

/// Whether an optional envelope field may appear.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Presence {
    /// Must be present.
    Required,
    /// May be present.
    Optional,
    /// Must be absent. A field with no meaning for an operation is a field that
    /// could carry an ambiguous meaning, so it is refused.
    Forbidden,
}

impl Presence {
    /// Wire spelling, as used in schema annotations.
    #[must_use]
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Required => "required",
            Self::Optional => "optional",
            Self::Forbidden => "forbidden",
        }
    }
}

/// Per-message presence rules for the optional envelope fields.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct EnvelopeRules {
    /// `correlation_id`
    pub correlation_id: Presence,
    /// `causation_id`
    pub causation_id: Presence,
    /// `session_id`
    pub session_id: Presence,
    /// `run_id`
    pub run_id: Presence,
    /// `epoch`
    pub epoch: Presence,
    /// `idempotency_key` — mandatory on any request with a side effect
    /// (`PROTOCOL.md` §1); no M2 operation has one.
    pub idempotency_key: Presence,
}

impl EnvelopeRules {
    /// Every optional field permitted, none required. Used to inspect messages
    /// whose schema the receiver does not know (DWCP, events).
    pub const PERMISSIVE: Self = Self {
        correlation_id: Presence::Optional,
        causation_id: Presence::Optional,
        session_id: Presence::Optional,
        run_id: Presence::Optional,
        epoch: Presence::Optional,
        idempotency_key: Presence::Optional,
    };

    /// The rules as `(field name, presence)` pairs, in decode order.
    #[must_use]
    pub const fn fields(&self) -> [(&'static str, Presence); 6] {
        [
            ("correlation_id", self.correlation_id),
            ("causation_id", self.causation_id),
            ("session_id", self.session_id),
            ("run_id", self.run_id),
            ("epoch", self.epoch),
            ("idempotency_key", self.idempotency_key),
        ]
    }
}

/// Every key the envelope declares.
pub const ENVELOPE_KEYS: [&str; 13] = [
    "v",
    "id",
    "type",
    "schema",
    "schema_version",
    "ts",
    "correlation_id",
    "causation_id",
    "session_id",
    "run_id",
    "epoch",
    "idempotency_key",
    "payload",
];

/// Enough of the envelope to identify a message: steps 1–3.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Preamble {
    /// Envelope version, already checked against the receiver's support.
    pub v: Version,
    /// Message type.
    pub message_type: MessageType,
    /// Message schema name.
    pub schema: SchemaName,
    /// Schema version (not yet checked against a registry).
    pub schema_version: Version,
}

/// The decoded envelope.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Header {
    /// Envelope version.
    pub v: Version,
    /// Message id; prefix `msg` for requests and responses, `evt` for events.
    pub id: AnyId,
    /// Message type.
    pub message_type: MessageType,
    /// Message schema name.
    pub schema: SchemaName,
    /// Schema version.
    pub schema_version: Version,
    /// Sender's timestamp. Advisory.
    pub ts: Timestamp,
    /// Groups the messages of one logical operation.
    pub correlation_id: Option<AnyId>,
    /// The message or event that directly caused this one.
    pub causation_id: Option<AnyId>,
    /// Session the message concerns. A claim the kernel verifies.
    pub session_id: Option<SessionId>,
    /// Run the message concerns. A claim the kernel verifies.
    pub run_id: Option<RunId>,
    /// Lease epoch the sender believes it holds. The kernel fences on it.
    pub epoch: Option<Epoch>,
    /// Idempotency key for side-effecting requests.
    pub idempotency_key: Option<IdempotencyKey>,
}

/// Steps 1–3: identify the message without interpreting anything else.
pub fn read_preamble(value: &Value, supported: VersionRange) -> Result<Preamble, ProtocolError> {
    let Value::Object(object) = value else {
        return Err(ProtocolError::schema(
            Violation::WrongType,
            "",
            format!("a message is an object, found {}", value.type_name()),
        ));
    };
    let mut cx = Cx::new();

    let v: Version = peek_required(object, "v", &mut cx)?;
    if !supported.contains(v.get()) {
        return Err(ProtocolError::new(
            crate::error::ErrorCode::VersionUnsupported,
            format!(
                "envelope version {} is not supported; supported {}..={}",
                v.get(),
                supported.min,
                supported.max
            ),
        )
        .with_violation(Violation::OutOfRange)
        .with_path("/v")
        .with_supported(supported));
    }
    let message_type: MessageType = peek_required(object, "type", &mut cx)?;
    let schema: SchemaName = peek_required(object, "schema", &mut cx)?;
    let schema_version: Version = peek_required(object, "schema_version", &mut cx)?;
    Ok(Preamble {
        v,
        message_type,
        schema,
        schema_version,
    })
}

fn peek_required<T: WireType>(
    object: &Object,
    name: &str,
    cx: &mut Cx,
) -> Result<T, ProtocolError> {
    cx.push(name);
    let result = match object.get(name) {
        None => {
            cx.pop();
            return Err(ProtocolError::schema(
                Violation::MissingField,
                "",
                format!("missing envelope field {}", quote_key(name)),
            )
            .with_path(&format!("/{name}")));
        }
        Some(Value::Null) => Err(cx.violation(Violation::NullNotAllowed, "null is not a value")),
        Some(value) => T::decode(value.clone(), cx),
    };
    cx.pop();
    result
}

/// Steps 4–7: decode the rest of the envelope against `rules`.
///
/// Returns the header, the raw payload object, and any preserved undeclared
/// envelope members (always empty under [`UnknownFields::Reject`]).
pub fn decode_header(
    value: Value,
    preamble: Preamble,
    rules: EnvelopeRules,
    unknown: UnknownFields,
) -> Result<(Header, Object, Object), ProtocolError> {
    let Value::Object(object) = value else {
        return Err(ProtocolError::schema(
            Violation::WrongType,
            "",
            "a message is an object",
        ));
    };
    let mut cx = Cx::new();
    let (mut members, extensions) = partition_members(object, &ENVELOPE_KEYS, unknown, &cx)?;

    let id: AnyId = take_required(&mut members, "id", &mut cx)?;
    let ts: Timestamp = take_required(&mut members, "ts", &mut cx)?;
    let correlation_id = take_ruled(
        &mut members,
        "correlation_id",
        rules.correlation_id,
        &mut cx,
    )?;
    let causation_id = take_ruled(&mut members, "causation_id", rules.causation_id, &mut cx)?;
    let session_id = take_ruled(&mut members, "session_id", rules.session_id, &mut cx)?;
    let run_id = take_ruled(&mut members, "run_id", rules.run_id, &mut cx)?;
    let epoch: Option<Epoch> = take_ruled(&mut members, "epoch", rules.epoch, &mut cx)?;
    let idempotency_key = take_ruled(
        &mut members,
        "idempotency_key",
        rules.idempotency_key,
        &mut cx,
    )?;

    if epoch.is_some() && session_id.is_none() {
        return Err(ProtocolError::schema(
            Violation::Inconsistent,
            "/epoch",
            "an epoch is meaningful only for a session; session_id is required with epoch",
        ));
    }
    let expected_prefix = preamble.message_type.id_prefix();
    if id.prefix() != expected_prefix {
        return Err(ProtocolError::schema(
            Violation::InvalidFormat,
            "/id",
            format!(
                "a {} uses a {expected_prefix}_ id",
                preamble.message_type.as_str()
            ),
        ));
    }

    let payload = match members.iter().position(|(k, _)| k == "payload") {
        None => {
            return Err(ProtocolError::schema(
                Violation::MissingField,
                "/payload",
                "missing envelope field \"payload\"",
            ));
        }
        Some(index) => members.remove(index).1,
    };
    let payload = match payload {
        Value::Object(object) => object,
        Value::Null => {
            return Err(ProtocolError::schema(
                Violation::NullNotAllowed,
                "/payload",
                "null is not a value",
            ));
        }
        other => {
            return Err(ProtocolError::schema(
                Violation::WrongType,
                "/payload",
                format!("expected object, found {}", other.type_name()),
            ));
        }
    };

    let header = Header {
        v: preamble.v,
        id,
        message_type: preamble.message_type,
        schema: preamble.schema,
        schema_version: preamble.schema_version,
        ts,
        correlation_id,
        causation_id,
        session_id,
        run_id,
        epoch,
        idempotency_key,
    };
    Ok((header, payload, extensions))
}

fn take_ruled<T: WireType>(
    members: &mut Vec<(String, Value)>,
    name: &str,
    presence: Presence,
    cx: &mut Cx,
) -> Result<Option<T>, ProtocolError> {
    let present = members.iter().any(|(k, _)| k == name);
    match (presence, present) {
        (Presence::Forbidden, true) => Err(ProtocolError::schema(
            Violation::ForbiddenField,
            &format!("/{name}"),
            format!(
                "envelope field {} is not permitted for this operation",
                quote_key(name)
            ),
        )),
        (Presence::Forbidden, false) => Ok(None),
        (Presence::Required, _) => take_required(members, name, cx).map(Some),
        (Presence::Optional, _) => take_optional(members, name, cx),
    }
}

impl Header {
    /// Encode the envelope around `payload`, re-emitting preserved members.
    pub fn encode(&self, payload: Value, extensions: &Object) -> Result<Value, ProtocolError> {
        let mut object = Object::new();
        object.insert("v".to_owned(), self.v.encode()?)?;
        object.insert("id".to_owned(), self.id.encode()?)?;
        object.insert("type".to_owned(), self.message_type.encode()?)?;
        object.insert("schema".to_owned(), self.schema.encode()?)?;
        object.insert("schema_version".to_owned(), self.schema_version.encode()?)?;
        object.insert("ts".to_owned(), self.ts.encode()?)?;
        put_optional(&mut object, "correlation_id", self.correlation_id.as_ref())?;
        put_optional(&mut object, "causation_id", self.causation_id.as_ref())?;
        put_optional(&mut object, "session_id", self.session_id.as_ref())?;
        put_optional(&mut object, "run_id", self.run_id.as_ref())?;
        put_optional(&mut object, "epoch", self.epoch.as_ref())?;
        put_optional(
            &mut object,
            "idempotency_key",
            self.idempotency_key.as_ref(),
        )?;
        object.insert("payload".to_owned(), payload)?;
        for (key, value) in extensions.iter() {
            object.insert(key.to_owned(), value.clone())?;
        }
        Ok(Value::Object(object))
    }
}

fn put_optional<T: WireType>(
    object: &mut Object,
    name: &str,
    value: Option<&T>,
) -> Result<(), ProtocolError> {
    match value {
        Some(v) => object.insert(name.to_owned(), v.encode()?),
        None => Ok(()),
    }
}

/// Schema nodes for the envelope fields, keyed by field name.
pub fn field_schema(name: &str, defs: &mut Defs) -> Option<Value> {
    Some(match name {
        "id" | "correlation_id" | "causation_id" => AnyId::schema(defs),
        "ts" => Timestamp::schema(defs),
        "session_id" => SessionId::schema(defs),
        "run_id" => RunId::schema(defs),
        "epoch" => Epoch::schema(defs),
        "idempotency_key" => IdempotencyKey::schema(defs),
        _ => return None,
    })
}
